src/apps/js_generic/js_generic_base.cpp

```diff
@@ -253,57 +253,86 @@ namespace ccfapp
+            js::ReadOnlyTxContext historical_txctx{&tx};
+            auto add_historical_globals = [&](js::Context& ctx) {
+              js::populate_global_ccf_historical_state(
+                &historical_txctx, tx_id, receipt, ctx);
+            };
+            do_execute_request(endpoint, endpoint_ctx, add_historical_globals);
           },
           context,
           is_tx_committed)(endpoint_ctx);
       }
       else
       {
-        do_execute_request(
-          endpoint, endpoint_ctx, nullptr, std::nullopt, nullptr);
+        do_execute_request(endpoint, endpoint_ctx);
       }
     }
 
+    using PreExecutionHook = std::function<void(js::Context&)>;
+
     void do_execute_request(
-      const JSDynamicEndpoint* endpoint,
+      const ccf::js::JSDynamicEndpoint* endpoint,
       ccf::endpoints::EndpointContext& endpoint_ctx,
-      kv::ReadOnlyTx* historical_tx,
-      const std::optional<ccf::TxID>& transaction_id,
-      ccf::TxReceiptImplPtr receipt)
+      const std::optional<PreExecutionHook>& pre_exec_hook = std::nullopt)
     {
-      js::Runtime rt(&endpoint_ctx.tx);
-      rt.add_ccf_classdefs();
-
+      // This KV Value should be updated by any governance actions which modify
+      // the JS app (including _any_ of its contained modules). We then use the
+      // version where it was last modified as a safe approximation of when an
+      // interpreter is unsafe to use. If this value is written to, the
+      // version_of_previous_write will advance, and all cached interpreters
+      // will be flushed.
+      const auto interpreter_flush = endpoint_ctx.tx.ro<ccf::InterpreterFlush>(
+        ccf::Tables::INTERPRETER_FLUSH);
+      const auto flush_marker =
+        interpreter_flush->get_version_of_previous_write().value_or(0);
+
+      const std::optional<JSRuntimeOptions> js_runtime_options =
+        endpoint_ctx.tx.ro<ccf::JSEngine>(ccf::Tables::JSENGINE)->get();
+      if (js_runtime_options.has_value())
+      {
+        interpreter_cache->set_max_cached_interpreters(
+          js_runtime_options->max_cached_interpreters);
+      }
+
+      std::shared_ptr<js::Context> interpreter =
+        interpreter_cache->get_interpreter(
+          js::TxAccess::APP, *endpoint, flush_marker);
+      if (interpreter == nullptr)
+      {
+        throw std::logic_error("Cache failed to produce interpreter");
+      }
+      js::Context& ctx = *interpreter;
+
+      // Prevent any other thread modifying this interpreter, until this
+      // function completes. We could create interpreters per-thread, but then
+      // we would get no cross-thread caching benefit (and would need to either
+      // enforce, or share, caps across per-thread caches). We choose
+      // instead to allow interpreters to be maximally reused, even across
+      // threads, at the cost of locking (and potentially stalling another
+      // thread's request execution) here.
```

Comment:
That may be something to revisit if we add an explicit way to populate the interpreters (ie. a specific "setup_global_state()" call in the application that CCF would call), in which case keeping one per thread is likely a good memory vs throughput tradeoff.

Reply:
I agree that we should revisit this in future, especially when memory use is no longer such a constraint. I think this is a sensible behaviour for now, and hopefully the comment reminds us to revisit.